### PR TITLE
add partition ID expression for sharding PK ID

### PR DESCRIPTION
**Pull request by IANTHEREAL**
add one expression for sharding PK ID to partition these sharding table.

Some simple instructions
* expression of rule is "partition id"
* arguments of rule contains [instance_id, prefix of schema, prefix of table]
   we would compute a ID like
   [1:1 bit][2:9 bits][3:10 bits][4:44 bits] int64  (using default bits length)
     * 1 useless, no reason
     * 2 schema ID (schema suffix)
     * 3 table ID (table suffix)
     * 4 origin ID (>= 0, <= 17592186044415)

NOTICE: others: schema = arguments[1] + schema suffix, table = arguments[2] + table suffix

### CODE REVIEW

File: pkg/binlog-filter/filter.go [hunk 1]
```diff
@@ -96,6 +96,10 @@ type BinlogEvent struct {
 
 // NewBinlogEvent returns a binlog event filter
 func NewBinlogEvent(rules []*BinlogEventRule) (*BinlogEvent, error) {
+	if len(rules) == 0 {
+		return nil, nil
+	}
+
```
Comment (csuzhangxc) [inline pkg/binlog-filter/filter.go]: so now, we can not create a Filter with no rules? (we can't split New and AddRule ?)

File: pkg/column-mapping/column_test.go [hunk 2]
```diff
@@ -45,22 +45,22 @@ func (t *testColumnMappingSuit) TestHandle(c *C) {
 	c.Assert(m.cache.infos, HasLen, 0)
 
 	// test clone
-	vals, err := m.HandleRowValue("test", "abc", []string{"id", "copyid", "name"}, []interface{}{1, "name"})
+	vals, _, err := m.HandleRowValue("test", "abc", []string{"id", "copyid", "name"}, []interface{}{1, "name"})
```
Comment (csuzhangxc) [inline pkg/column-mapping/column_test.go]: do assert on returned `[]int` ?

File: pkg/binlog-filter/filter.go (hunk 1 above)
Comment (IANTHEREAL) [reply pkg/binlog-filter/filter.go]: review it again after I finish it

File: pkg/column-mapping/column_test.go (hunk 2 above)
Comment (IANTHEREAL) [reply pkg/column-mapping/column_test.go]: it's a very trival case, bot are ok

**Comment by IANTHEREAL** [comment]
@csuzhangxc PTAL

File: pkg/column-mapping/column.go [hunk 3]
```diff
@@ -302,23 +366,119 @@ func tableName(schema, table string) string {
...
+		vals[info.targetPosition] = originID
+	}
+
+	return vals, nil
 }
 
-func cloneColumn(info *columnInfo, vals []interface{}) []interface{} {
-	newVals := make([]interface{}, 0, len(vals)+1)
-	newVals = append(newVals, vals[:info.targetPosition]...)
-	newVals = append(newVals, vals[info.sourcePosition])
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go:320]: seems `sourcePosition` is not used anymore now. keep it for future usage?

File: pkg/column-mapping/README.md [hunk 4]
```diff
@@ -4,11 +4,11 @@
 
 column mapping is a library to provide a simple and unified way to mapping columns of table:
 
-- add prefix for one column
+- add prefix for one char/varchar/ttext column
```
Comment (csuzhangxc) [inline pkg/column-mapping/README.md]: s/ttext/text

File: pkg/column-mapping/column.go [hunk 5]
```diff
@@ -15,31 +15,59 @@ package column
...
 
 	"github.com/juju/errors"
 	"github.com/pingcap/tidb-tools/pkg/table-rule-selector"
 )
 
+var (
+	// for partition ID, ref definition of partitionID
+	instanceIDBitSize       = 4
+	schemaIDBitSize         = 7
+	tableIDBitSize          = 8
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go:29]: define them as `uint` ?

File: pkg/column-mapping/README.md [hunk 6]
```diff
@@ -29,11 +29,19 @@ type Rule struct {
...
+
+add suffix, with arguments[suffix]
+
+partition id, with arguments [instance_id, prefix of schema, prefix of table]
+[1:1 bit][2:9 bits][3:10 bits][4:44 bits] int64  (using default bits length)
+- 1: useless, no reason
+- 2: schema ID (schema suffix)
+- 3: table ID (table suffix)
+- 4: origin ID (>= 0, <= 17592186044415)
+And schema = arguments[1] + schema suffix, table = arguments[2] + table suffix
```
Comment (csuzhangxc) [inline pkg/column-mapping/README.md:42]: I didn't understand this until I read `computeID` :joy:
maybe you can give an example

File: pkg/column-mapping/column.go [hunk 7]
```diff
@@ -302,23 +366,119 @@ func tableName(schema, table string) string {
...
+
+func computeID(name string, prefix string, bitSize int, shiftCount uint) (int64, error) {
+	if len(prefix) >= len(name) || prefix != name[:len(prefix)] {
+		return 0, errors.NotValidf("%s is not prefix of %s", prefix, name)
+	}
+
+	idStr := name[len(prefix):]
+	id, err := strconv.ParseUint(idStr, 10, bitSize)
+	if err != nil {
+		return 0, errors.NotValidf("suffix of %d is not int64", idStr)
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go]: s/%d/%s

File: pkg/column-mapping/column.go [hunk 8]
```diff
@@ -15,31 +15,59 @@ package column
...
+	tableIDBitSize          = 8
+	maxOriginID       int64 = 17592186044416
+)
+
+// SetPartitionRule sets bit size of schema ID and table ID
+func SetPartitionRule(instanceIDSize, schemaIDSize, tableIDSize int) {
+	instanceIDBitSize = instanceIDSize
+	schemaIDBitSize = schemaIDSize
+	tableIDBitSize = tableIDSize
+	maxOriginID = 1 << uint(64-instanceIDSize-schemaIDSize-tableIDSize-1)
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go:38]: do we need to check the total size of `instanceIDSize`, `schemaIDBitSize`, `tableIDBitSize`

File: pkg/column-mapping/column.go [hunk 9]
```diff
@@ -302,23 +366,119 @@ func tableName(schema, table string) string {
...
+		originID = rawID
+	case uint:
+		originID = int64(rawID)
+	case uint16:
+		originID = int64(rawID)
+	case uint32:
+		originID = int64(rawID)
+	case uint64:
+		originID = int64(rawID)
+	case string:
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go:427]: a `default` clause needed?

File: pkg/column-mapping/column.go [hunk 10]
```diff
@@ -302,23 +366,119 @@ func tableName(schema, table string) string {
...
 
-func cloneColumn(info *columnInfo, vals []interface{}) []interface{} {
-	newVals := make([]interface{}, 0, len(vals)+1)
-	newVals = append(newVals, vals[:info.targetPosition]...)
-	newVals = append(newVals, vals[info.sourcePosition])
-	newVals = append(newVals, vals[info.targetPosition:]...)
+func computePartitionID(schema, table string, rule *Rule) (instanceID int64, schemaID int64, tableID int64, err error) {
+	shiftCnt := uint(64 - instanceIDBitSize - 1)
+	if instanceIDBitSize > 0 {
+		instanceID, err = strconv.ParseInt(rule.Arguments[0], 10, 64)
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go]: `strconv.ParseInt(rule.Arguments[0], 10, instanceIDBitSize)` ?

File: pkg/column-mapping/column_test.go [hunk 11]
```diff
@@ -27,40 +28,172 @@ var _ = Suite(&testColumnMappingSuit{})
...
+	}
+
 	// initial column mapping
 	m, err := NewMapping(rules)
 	c.Assert(err, IsNil)
 	c.Assert(m.cache.infos, HasLen, 0)
 
 	// test clone
-	vals, err := m.HandleRowValue("test", "abc", []string{"id", "copyid", "name"}, []interface{}{1, "name"})
+	vals, poss, err := m.HandleRowValue("test", "xxx", []string{"age", "id"}, []interface{}{1, "1"})
```
Comment (csuzhangxc) [inline pkg/column-mapping/column_test.go]: no `test clone` exist anymore

File: pkg/column-mapping/column.go (hunk 3 above)
Comment (IANTHEREAL) [reply pkg/column-mapping/column.go:320]: yep

File: pkg/column-mapping/column.go (hunk 5 above)
Comment (IANTHEREAL) [reply pkg/column-mapping/column.go:29]: it's ok, like `strconv.ParseUint`, bitsize is int

File: pkg/column-mapping/column_test.go (hunk 11 above)
Comment (IANTHEREAL) [reply pkg/column-mapping/column_test.go]: not `clone` now

File: pkg/column-mapping/column.go (hunk 8 above)
Comment (IANTHEREAL) [reply pkg/column-mapping/column.go:38]: that's all, but there are more restrictions later

**Comment by IANTHEREAL** [comment]
after #58  merged, ci would be fixed

File: pkg/column-mapping/column.go [hunk 12]
```diff
@@ -302,23 +368,122 @@ func tableName(schema, table string) string {
...
+	case uint64:
+		originID = int64(rawID)
+	case string:
+		originID, err = strconv.ParseInt(rawID, 10, 64)
+		if err != nil {
+			return nil, errors.NotValidf("column %d value is not int, but %v", info.targetPosition, vals[info.targetPosition])
+		}
+		isChars = true
+	default:
+		return nil, errors.NotValidf("type %T(%v)", vals[info.targetPosition])
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go]: seems `%T` has no arg

File: pkg/column-mapping/column_test.go (hunk 11 above)
Comment (csuzhangxc) [reply pkg/column-mapping/column_test.go]: so change to `// test add prefix` ?

**Comment by csuzhangxc** [comment]
rest LGTM

**Pull request merged**